### PR TITLE
BMC: Added CLI commands runner and Serial Console access.

### PR DESCRIPTION
**Pull request by greyerof**
BMC's CLI commands can be run using the `bmc.RunCLICommand` method, which connects to the CLI via SSH. The method blocks until the command ends or a timeout happens, whatever comes first. The command's output is copied into the stdout & stderr string vars. The underlying SSH session is created and closed internally after the command has finished.

The new method `bmc.OpenSerialConsole` returns a piped reader and writer interfaces that can be used to interactively receive/send commands/text to/from the serial console. Since the serial console is tunneled in a SSH session, the `bmc.CloseSerialConsole` method must be called to release that session to avoid leaks that would prevent CLI commands to fail due to maximum concurrent SSH sessions reached in the BMC.

Also, added SSH port as parameter for BMC's New() method

### CODE REVIEW

File: pkg/bmc/bmc.go [hunk 1]
```diff
@@ -44,14 +56,17 @@ type BMC struct {
...
 
 	timeOuts TimeOuts
+
+	sshSessionForSerialConsole *ssh.Session
 }
 
 // New returns a new BMC struct. The default system index to be used in redfish requests is 0.
 // Use SetSystemIndex to modify it.
-func New(host string, redfishUser, sshUser User, timeOuts TimeOuts) (*BMC, error) {
+func New(host string, redfishUser, sshUser User, sshPort int, timeOuts TimeOuts) (*BMC, error) {
```
Comment (kononovn) [inline pkg/bmc/bmc.go:69]: May be ssh port has to be uint instead of int?

File: pkg/bmc/bmc.go [hunk 2]
```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+		return nil, nil, fmt.Errorf("there is already a serial console opened for this BMC")
+	}
+
+	cliCmd := openConsoleCliCmd
+	if cliCmd == "" {
+		// no cli command to get console port was provided, try to guess based on
+		// manufacturer.
+		manufacturer, err := bmc.SystemManufacturer()
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get redfish system manufacturer: %w", err)
```
Comment (kononovn) [inline pkg/bmc/bmc.go:486]: Please add glog

File: pkg/bmc/bmc.go [hunk 3]
```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+		}
+
+		var found bool
+		if cliCmd, found = cliCmdSerialConsole[manufacturer]; !found {
+			return nil, nil, fmt.Errorf("cli command to get serial console not found for manufacturer %v", manufacturer)
```
Comment (kononovn) [inline pkg/bmc/bmc.go:491]: Please add glog

File: pkg/bmc/bmc.go [hunk 4]
```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+		}
+	}
+
+	sshSession, err := bmc.CreateCLISSHSession()
+	if err != nil {
```
Comment (kononovn) [inline pkg/bmc/bmc.go:496]: Please add glog

File: pkg/bmc/bmc.go [hunk 5]
```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+		return nil, nil, err
+	}
+
+	// Pipes need to be retrieved before session.Start()
+	reader, err := sshSession.StdoutPipe()
+	if err != nil {
```
Comment (kononovn) [inline pkg/bmc/bmc.go:502]: Please add glog

File: pkg/bmc/bmc.go [hunk 6]
```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to get stdout pipe from ssh session: %w", err)
+	}
+
+	writer, err := sshSession.StdinPipe()
+	if err != nil {
```
Comment (kononovn) [inline pkg/bmc/bmc.go:509]: Please add glog

File: pkg/bmc/bmc.go [hunk 7]
```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to get stdin pipe from ssh session: %w", err)
+	}
+
+	err = sshSession.Start(cliCmd)
+	if err != nil {
```
Comment (kononovn) [inline pkg/bmc/bmc.go:516]: Please add glog

File: pkg/bmc/bmc.go [hunk 8]
```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+}
+
+// CloseSerialConsole closes the serial console's underlying ssh session.
+func (bmc *BMC) CloseSerialConsole() error {
+	if bmc.sshSessionForSerialConsole == nil {
+		return fmt.Errorf("no underlying ssh session found")
+	}
+
+	err := bmc.sshSessionForSerialConsole.Close()
+	if err != nil {
```
Comment (kononovn) [inline pkg/bmc/bmc.go:536]: Please add glog

**Review by kononovn — CHANGES_REQUESTED** [review]
Few minor comments to fix

File: pkg/bmc/bmc.go (hunk 2 above)
Comment (greyerof) [reply pkg/bmc/bmc.go:492]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

File: pkg/bmc/bmc.go (hunk 3 above)
Comment (greyerof) [reply pkg/bmc/bmc.go:499]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

File: pkg/bmc/bmc.go (hunk 4 above)
Comment (greyerof) [reply pkg/bmc/bmc.go:504]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

File: pkg/bmc/bmc.go (hunk 5 above)
Comment (greyerof) [reply pkg/bmc/bmc.go:512]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

File: pkg/bmc/bmc.go (hunk 6 above)
Comment (greyerof) [reply pkg/bmc/bmc.go:521]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

File: pkg/bmc/bmc.go (hunk 7 above)
Comment (greyerof) [reply pkg/bmc/bmc.go:530]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

File: pkg/bmc/bmc.go (hunk 8 above)
Comment (greyerof) [reply pkg/bmc/bmc.go:556]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

File: pkg/bmc/bmc.go (hunk 1 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e

**Review by kononovn — DISMISSED** [review]
/lgtm

File: pkg/bmc/bmc.go [hunk 9]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+			return stdoutBuffer.String(), stderrBuffer.String(), fmt.Errorf("command run error: %w", err)
+		}
+	}
+
+	if combineOutput {
+		return string(combinedOutput), "", nil
+	}
+
+	return stdoutBuffer.String(), stderrBuffer.String(), nil
+}
```
Comment (trewest) [inline pkg/bmc/bmc.go:469]: @kononovn These operations don't seem to be specific to BMC, do you see an issue with adding them to a more generic pkg?

File: pkg/bmc/bmc.go [hunk 10]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+// OpenSerialConsole opens the serial console port. The console is tunneled in an underlying (CLI)
+// ssh session that is opened in the BMC's ssh server. If openConsoleCliCmd is
+// provided, it will be sent to the BMC's cli. Otherwise, a best effort will
+// be made to run the appropriate cli command based on the system manufacturer.
+func (bmc *BMC) OpenSerialConsole(openConsoleCliCmd string) (io.Reader, io.WriteCloser, error) {
+	glog.V(100).Infof("Opening serial console.")
```
Comment (trewest) [inline pkg/bmc/bmc.go:476]:
```suggestion
	glog.V(100).Infof("Opening serial console on %s.", bmc.host)
```

File: pkg/bmc/bmc.go [hunk 11]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+	if bmc.sshSessionForSerialConsole != nil {
+		glog.V(100).Infof("There is already a serial console opened for this BMC. Use OpenSerialConsole() first.")
```
Comment (trewest) [inline pkg/bmc/bmc.go:479]:
```suggestion
		glog.V(100).Infof("There is already a serial console opened for %s's BMC. Use OpenSerialConsole() first.", bmc.host)
```

File: pkg/bmc/bmc.go [hunk 12]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+		if cliCmd, found = cliCmdSerialConsole[manufacturer]; !found {
+			glog.V(100).Infof("CLI command to get serial console not found for manufacturer %v", manufacturer)
+
+			return nil, nil, fmt.Errorf("cli command to get serial console not found for manufacturer %v", manufacturer)
+		}
+	}
+
+	sshSession, err := bmc.CreateCLISSHSession()
+	if err != nil {
+		glog.V(100).Infof("Failed to create underlying ssh session: %v", err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:505]:
```suggestion
		glog.V(100).Infof("Failed to create underlying ssh session for %s: %v", bmc.host, err)
```

File: pkg/bmc/bmc.go [hunk 13]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+	}
+
+	// Pipes need to be retrieved before session.Start()
+	reader, err := sshSession.StdoutPipe()
+	if err != nil {
+		glog.V(100).Infof("Failed to get stdout pipe from ssh session: %v", err)
+
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to get stdout pipe from ssh session: %w", err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:517]:
```suggestion
		return nil, nil, fmt.Errorf("failed to get stdout pipe from %s's ssh session: %w", bmc.host, err)
```

File: pkg/bmc/bmc.go [hunk 14]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+	}
+
+	writer, err := sshSession.StdinPipe()
+	if err != nil {
+		glog.V(100).Infof("Failed to get stdin pipe from ssh session: %v", err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:522]:
```suggestion
		glog.V(100).Infof("Failed to get stdin pipe from %s's ssh session: %v", bmc.host, err)
```

File: pkg/bmc/bmc.go [hunk 15]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to get stdin pipe from ssh session: %w", err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:526]:
```suggestion
		return nil, nil, fmt.Errorf("failed to get stdin pipe from %s's ssh session: %w", bmc.host, err)
```

File: pkg/bmc/bmc.go [hunk 16]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+	}
+
+	err = sshSession.Start(cliCmd)
+	if err != nil {
+		glog.V(100).Infof("Failed to start CLI command %q: %v", cliCmd, err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:531]:
```suggestion
		glog.V(100).Infof("Failed to start CLI command %q on %s: %v", cliCmd, bmc.host, err)
```

File: pkg/bmc/bmc.go [hunk 17]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+		_ = sshSession.Close()
+
+		return nil, nil, fmt.Errorf("failed to start serial console with cli command %q: %w", cliCmd, err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:535]:
```suggestion
		return nil, nil, fmt.Errorf("failed to start serial console with cli command %q on %s: %w", cliCmd, bmc.host, err)
```

File: pkg/bmc/bmc.go [hunk 18]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+	return reader, writer, nil
+}
+
+// CloseSerialConsole closes the serial console's underlying ssh session.
+func (bmc *BMC) CloseSerialConsole() error {
+	glog.V(100).Infof("Closing serial console.")
+
+	if bmc.sshSessionForSerialConsole == nil {
+		glog.V(100).Infof("No underlying ssh session found. Please use OpenSerialConsole() first.")
```
Comment (trewest) [inline pkg/bmc/bmc.go:550]:
```suggestion
		glog.V(100).Infof("No underlying ssh session found for %s. Please use OpenSerialConsole() first.", bmc.host)
```

File: pkg/bmc/bmc.go [hunk 19]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+	go func() { _ = sshSession.Wait() }()
+
+	bmc.sshSessionForSerialConsole = sshSession
+
+	return reader, writer, nil
+}
+
+// CloseSerialConsole closes the serial console's underlying ssh session.
+func (bmc *BMC) CloseSerialConsole() error {
+	glog.V(100).Infof("Closing serial console.")
```
Comment (trewest) [inline pkg/bmc/bmc.go:547]:
```suggestion
	glog.V(100).Infof("Closing serial console on %s.", bmc.host)
```

File: pkg/bmc/bmc.go [hunk 20]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+	if bmc.sshSessionForSerialConsole == nil {
+		glog.V(100).Infof("No underlying ssh session found. Please use OpenSerialConsole() first.")
+
+		return fmt.Errorf("no underlying ssh session found")
```
Comment (trewest) [inline pkg/bmc/bmc.go:552]:
```suggestion
		return fmt.Errorf("no underlying ssh session found for %s", bmc.host)
```

File: pkg/bmc/bmc.go [hunk 21]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+	}
+
+	err := bmc.sshSessionForSerialConsole.Close()
+	if err != nil {
+		glog.V(100).Infof("Failed to close underlying ssh session: %v", err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:557]:
```suggestion
		glog.V(100).Infof("Failed to close underlying ssh session on %s: %v", bmc.host, err)
```

File: pkg/bmc/bmc.go [hunk 22]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+		return fmt.Errorf("failed to close underlying ssh session: %w", err)
```
Comment (trewest) [inline pkg/bmc/bmc.go:559]:
```suggestion
		return fmt.Errorf("failed to close underlying ssh session on %s: %w", bmc.host, err)
```

File: pkg/bmc/bmc.go (hunk 10 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 11 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go [hunk 23]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+// ssh session that is opened in the BMC's ssh server. If openConsoleCliCmd is
+// provided, it will be sent to the BMC's cli. Otherwise, a best effort will
+// be made to run the appropriate cli command based on the system manufacturer.
+func (bmc *BMC) OpenSerialConsole(openConsoleCliCmd string) (io.Reader, io.WriteCloser, error) {
+	glog.V(100).Infof("Opening serial console.")
+
+	if bmc.sshSessionForSerialConsole != nil {
+		glog.V(100).Infof("There is already a serial console opened for this BMC. Use OpenSerialConsole() first.")
+
+		return nil, nil, fmt.Errorf("there is already a serial console opened for this BMC")
```
[the remark this answers is not in the document]
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go [hunk 24]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+	}
+
+	cliCmd := openConsoleCliCmd
+	if cliCmd == "" {
+		// no cli command to get console port was provided, try to guess based on
+		// manufacturer.
+		manufacturer, err := bmc.SystemManufacturer()
+		if err != nil {
+			glog.V(100).Infof("Failed to get redifsh system manufacturer: %v", err)
```
[the remark this answers is not in the document]
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go [hunk 25]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+
+			return nil, nil, fmt.Errorf("failed to get redfish system manufacturer: %w", err)
```
[the remark this answers is not in the document]
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 12 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go [hunk 26]
```diff
@@ -347,3 +367,199 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
...
+	if err != nil {
+		glog.V(100).Infof("Failed to create underlying ssh session: %v", err)
+
+		return nil, nil, err
+	}
+
+	// Pipes need to be retrieved before session.Start()
+	reader, err := sshSession.StdoutPipe()
+	if err != nil {
+		glog.V(100).Infof("Failed to get stdout pipe from ssh session: %v", err)
```
[the remark this answers is not in the document]
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 13 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 14 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 15 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 16 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 17 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 19 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 18 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 20 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 21 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 22 above)
Comment (greyerof) [reply pkg/bmc/bmc.go]: Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/c9a7fff0339cd048ad3ab836a5afd4b0cbb6fdb7

File: pkg/bmc/bmc.go (hunk 9 above)
Comment (kononovn) [reply pkg/bmc/bmc.go:469]: @trewest Actually we open ssh connection to bmc and them use bmc cmd in order to get access to OS console. I think it should be part of BMC package. WDYT ?

Comment (trewest) [reply pkg/bmc/bmc.go:469]: Resolving. I'll check interest in standalone SSH pkg with contributors and refactor this later if necessary

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**